### PR TITLE
testgrid upgrade support

### PR DESCRIPTION
**Pull request by laverya**

### CODE REVIEW

File: testgrid/tgapi/pkg/handlers/dequeue_instance.go [hunk 1]
```diff
@@ -13,18 +15,31 @@ type DequeueInstanceResponse struct {
...
 	testInstance, err := testinstance.GetNextEnqueued()
 	if err != nil {
-		testInstance, err = testinstance.GetOldEnqueued()
-		if err != nil {
-			JSON(w, 200, []DequeueInstanceResponse{})
+		if err != sql.ErrNoRows {
+			JSON(w, 500, []DequeueInstanceResponse{})
+			fmt.Printf("error dequeing instance: %s\n", err.Error())
 			return
+		} else {
```
Comment (emosbaugh) [inline testgrid/tgapi/pkg/handlers/dequeue_instance.go:31]: no need for else here

File: testgrid/tgapi/pkg/handlers/dequeue_instance.go [hunk 2]
```diff
@@ -13,18 +15,31 @@ type DequeueInstanceResponse struct {
...
+			testInstance, err = testinstance.GetOldEnqueued()
+			if err != nil {
+				if err != sql.ErrNoRows {
+					JSON(w, 500, []DequeueInstanceResponse{})
+					fmt.Printf("error dequeing old instance: %s\n", err.Error())
+					return
+				} else {
```
Comment (emosbaugh) [inline testgrid/tgapi/pkg/handlers/dequeue_instance.go:38]: no need for else here

File: testgrid/tgapi/pkg/handlers/dequeue_instance.go [hunk 3]
```diff
@@ -13,18 +15,31 @@ type DequeueInstanceResponse struct {
...
 }
 
 func DequeueInstance(w http.ResponseWriter, r *http.Request) {
 	testInstance, err := testinstance.GetNextEnqueued()
 	if err != nil {
-		testInstance, err = testinstance.GetOldEnqueued()
-		if err != nil {
-			JSON(w, 200, []DequeueInstanceResponse{})
+		if err != sql.ErrNoRows {
+			JSON(w, 500, []DequeueInstanceResponse{})
```
Comment (emosbaugh) [inline testgrid/tgapi/pkg/handlers/dequeue_instance.go:28]: why json?

File: testgrid/tgapi/pkg/handlers/dequeue_instance.go [hunk 4]
```diff
@@ -13,18 +15,31 @@ type DequeueInstanceResponse struct {
...
+			fmt.Printf("error dequeing instance: %s\n", err.Error())
 			return
+		} else {
+			testInstance, err = testinstance.GetOldEnqueued()
+			if err != nil {
+				if err != sql.ErrNoRows {
+					JSON(w, 500, []DequeueInstanceResponse{})
```
Comment (emosbaugh) [inline testgrid/tgapi/pkg/handlers/dequeue_instance.go:35]: why json?

**Review by salahalsaleh — APPROVED** [review]

File: testgrid/web/src/components/views/InstanceTable.jsx [hunk 5]
```diff
@@ -216,7 +226,15 @@ export default class InstanceTable extends React.Component {
     const rows = Object.keys(this.props.instancesMap).map(kurlURL => {
       return (
         <tr key={kurlURL}>
-          <td><span className="url" onClick={() => this.viewInstanceInstaller(this.props.instancesMap[kurlURL][0])}>{kurlURL}</span></td>
+          <td>
+            <span className="url" onClick={() => this.viewInstanceInstaller(this.props.instancesMap[kurlURL][0])}>{kurlURL}</span>
+            {(this.props.instancesMap[kurlURL][0].upgradeURL && this.props.instancesMap[kurlURL][0].upgradeURL !== "") &&
```
Comment (emosbaugh) [inline testgrid/web/src/components/views/InstanceTable.jsx:231]: im on js expert but i think this is redundant

**Review by emosbaugh — APPROVED** [review]

File: testgrid/tgapi/pkg/handlers/dequeue_instance.go (hunk 3 above)
Comment (laverya) [reply testgrid/tgapi/pkg/handlers/dequeue_instance.go:28]: It probably doesn't need to be

File: testgrid/web/src/components/views/InstanceTable.jsx [hunk 6]
```diff
@@ -216,7 +226,15 @@ export default class InstanceTable extends React.Component {
     const rows = Object.keys(this.props.instancesMap).map(kurlURL => {
       return (
         <tr key={kurlURL}>
-          <td><span className="url" onClick={() => this.viewInstanceInstaller(this.props.instancesMap[kurlURL][0])}>{kurlURL}</span></td>
+          <td>
+            <span className="url" onClick={() => this.viewInstanceInstaller(this.props.instancesMap[kurlURL][0])}>{kurlURL}</span>
+            {(this.props.instancesMap[kurlURL][0].upgradeURL !== "") &&
```
Comment (emosbaugh) [inline testgrid/web/src/components/views/InstanceTable.jsx:231]: i think this should be the following. @salahalsaleh can you help out here?

```
(this.props.instancesMap[kurlURL][0].upgradeURL) &&
```

**Review by emosbaugh — CHANGES_REQUESTED** [review]

File: testgrid/web/src/components/views/InstanceTable.jsx (hunk 5 above)
Comment (salahalsaleh) [reply testgrid/web/src/components/views/InstanceTable.jsx]: true, was going to comment on that, but it doesn't hurt

File: testgrid/web/src/components/views/InstanceTable.jsx (hunk 6 above)
Comment (laverya) [reply testgrid/web/src/components/views/InstanceTable.jsx:231]: I can do that, and yeah makes sense

Comment (salahalsaleh) [reply testgrid/web/src/components/views/InstanceTable.jsx]: yes you're right @emosbaugh, cuz that will also check for `null` & `undefined`.

Comment (salahalsaleh) [reply testgrid/web/src/components/views/InstanceTable.jsx]: also, no need for the parentheses

**Review by salahalsaleh — APPROVED** [review]

**Review by emosbaugh — APPROVED** [review]

**Pull request merged**